### PR TITLE
refactor dataset,  add channel, fix dataset bug

### PR DESCRIPTION
**Pull request by xjqbest**
(1) use channel instead of vector/BlockingQueue in Dataset，to keep same with existing implementation, and make code more readable and flexible (dataset single output channel or multi output channel). one previous memory out of limit problem is cause by not release memory after training.
(2) add Record because MultiSlotType costs too much memory (80B)，fix memory out of limit problem.
(3) add Channel, Archive in paddle/fluid/framework
(4) change dataset from shared_ptr to unique_ptr in pybind
(5) move create/destroy readers from trainer to dataset
(6) move shuffle from datafeed to dataset.  dataset holds memory, datafeed is only for load data and feed data to network.
(7) fix thread num bug of Dataset when filelist size < thread num
(8) support set_queue_num in InMemoryDataset

### CODE REVIEW

File: paddle/fluid/framework/archive.h [hunk 1]
```diff
@@ -0,0 +1,609 @@
...
+    deleter_ = nullptr;
+    return buf;
+  }
+
+  void Resize(size_t newsize) {
+#ifdef _LINUX
+    if (unlikely(newsize > Capacity())) {
+#else
+    if (newsize > Capacity()) {
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/archive.h:170]: 这里如果写成统一的形式，速度差别会很大吗？

File: paddle/fluid/framework/archive.h [hunk 2]
```diff
@@ -0,0 +1,609 @@
...
+      deleter_ = std::default_delete<char[]>();
+    }
+  }
+
+  void PrepareRead(size_t size) {
+#ifdef _LINUX
+    if (unlikely(!(size <= size_t(finish_ - cursor_)))) {
+#else
+    if (!(size <= size_t(finish_ - cursor_))) {
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/archive.h:199]: 同上

File: paddle/fluid/framework/archive.h [hunk 3]
```diff
@@ -0,0 +1,609 @@
...
+  return ar;
+}
+
+template <class AR, class T>
+Archive<AR>& operator<<(Archive<AR>& ar, const std::vector<T>& p) {
+#ifdef _LINUX
+  ar << (size_t)p.size();
+#else
+  ar << (uint64_t)p.size();
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/archive.h:338]: 为什么不写成统一形式呢？

File: paddle/fluid/framework/archive.h [hunk 4]
```diff
@@ -0,0 +1,609 @@
...
+  return ar;
+}
+
+template <class AR, class T>
+Archive<AR>& operator>>(Archive<AR>& ar, std::vector<T>& p) {
+#ifdef _LINUX
+  p.resize(ar.template Get<size_t>());
+#else
+  p.resize(ar.template Get<uint64_t>());
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/archive.h:351]: 同上

File: paddle/fluid/framework/archive.h [hunk 5]
```diff
@@ -0,0 +1,609 @@
...
+  return ar;
+}
+#else
+template <class AR, class... T>
+Archive<AR>& SerializeTuple(Archive<AR>& ar,                          // NOLINT
+                            const std::tuple<T...>& x,                // NOLINT
+                            std::integral_constant<uint64_t, 0> n) {  // NOLINT
+  return ar;
+}
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/archive.h:430]: 同上

File: paddle/fluid/framework/channel.h [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+  bool EmptyUnlocked() { return data_.empty(); }
+
+  bool FullUnlocked() { return data_.size() >= capacity_ + reading_count_; }
+
+  bool WaitForRead(std::unique_lock<std::mutex>& lock) {  // NOLINT
+#ifdef _LINUX
+    while (unlikely(EmptyUnlocked() && !closed_)) {
+#else
+    while (EmptyUnlocked() && !closed_) {
+#endif
```
Comment (chengduoZH) [inline paddle/fluid/framework/channel.h:216]: 同上

File: paddle/fluid/framework/archive.h (hunk 3 above)
Comment (xjqbest) [reply paddle/fluid/framework/archive.h:338]: mac/windows下这块用size_t编译会不通过，所以用了uint64_t。

File: paddle/fluid/framework/archive.h (hunk 1 above)
Comment (xjqbest) [reply paddle/fluid/framework/archive.h:170]: 速度还未专门测试过这块，不过此处加unlikely是跟已有的实现（内部版本）一致的。

**Review by jacquesqiao — APPROVED** [review]
LGTM!

**Pull request merged**